nomad/data_source_jwks.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package nomad
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"math/big"
+
+	"github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/id"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceJWKS() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceJWKSRead,
+		Schema: map[string]*schema.Schema{
+			"keys": {
+				Description: "JSON Web Key Set (JWKS) public keys for validating workload identity JWTs",
+				Type:        schema.TypeList,
+				Computed:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key_use": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"key_type": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"key_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"algorithm": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"modulus": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"exponent": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+			"pem_keys": {
+				Description: "JWKS as a list of PEM keys",
+				Type:        schema.TypeList,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Computed:    true,
+			},
+		},
+	}
+}
+
+type Key struct {
+	KeyUse    string `json:"use"`
+	KeyType   string `json:"kty"`
+	KeyId     string `json:"kid"`
+	Algorithm string `json:"alg"`
+	Modulus   string `json:"n"`
+	Exponent  string `json:"e"`
+}
+
+func dataSourceJWKSRead(d *schema.ResourceData, meta any) error {
+	client := meta.(ProviderConfig).client
+	operator := client.Raw()
+	queryOpts := &api.QueryOptions{}
+
+	jwks := struct {
+		Keys []Key `json:"keys"`
+	}{}
+
+	log.Printf("[DEBUG] Reading JWKS from Nomad")
+	_, err := operator.Query("/.well-known/jwks.json", &jwks, queryOpts)
+
+	if err != nil {
+		return fmt.Errorf("error reading JWKS from Nomad: %s", err)
+	}
+
+	if len(jwks.Keys) == 0 {
+		return fmt.Errorf("no keys found")
+	}
+
+	d.SetId(id.UniqueId())
+	if err := d.Set("keys", fromKeys(jwks.Keys)); err != nil {
+		return fmt.Errorf("error setting JWKS: %#v", err)
+	}
+
+	pemKeys := make([]string, 0, len(jwks.Keys))
+
+	for _, key := range jwks.Keys {
+		pemKey, err := keyToPem(key)
+		if err != nil {
+			return fmt.Errorf("Could not encode JWK as PEM: %s", err)
+		}
+		pemKeys = append(pemKeys, pemKey)
+	}
+
+	if err := d.Set("pem_keys", pemKeys); err != nil {
+		return fmt.Errorf("error setting JWKS pemKeys: %s", err)
+	}
+
+	return nil
+}
+
+func keyToPem(key Key) (string, error) {
+
+	// FIXME does Nomad always use RSA keys for JWKS?
```

Comment:
I'm not sure about this, but I think Nomad only uses RSA for this?

Reply:
In theory the root key can also use `EdDSA`, but I think for OIDC RSA is required.
https://github.com/hashicorp/nomad/blob/78f9f178670097eac9036c4bf83f7e0ceb55e6f3/nomad/encrypter.go#L499-L505

But we can probably remove this check altogether? It would also help future-proof this data source in case new algorithms are added later on.

Reply:
I did a bit of investigation on this: to support EdDSA it looks like we would need to add `crv` and `x` fields for `Key`s, and then construct a `ed25519` public key from them. We would still need the `KeyType` check to see which type of key to construct ("OKP" for EdDSA).

Also, as you pointed out, OIDC doesn't support non-RSA keys, so Nomad doesn't have code to make JWKs out of them. We would therefore need to construct a mocked-up key to test with.

I'm happy to do this, just want to make sure this sounds like worthwhile effort at this point. An alternative approach for future-proofing might be to ask for Nomad's `ListKeys` RPC to be make available via the HTTP API and fetch the keyset directly.

What do you think?

Reply:
Ah, I see. I think we can just leave RSA support like it is then, maybe just update the comment to explain why only RSA is supported?

Supporting new key types seem like will require code changes here as well, so not a bit deal.

Reply:
Looks like you and @wraithm already added a comment - thank you! I pushed a quick fix to change the indentation to match the surrounding code, I think it looks good to go now?